### PR TITLE
Add Maphyspp package and two dependencies : fabulous and paddle.

### PR DESCRIPTION
**Pull request by fpruvost**
See https://gitlab.inria.fr/solverstack/maphys/maphyspp

### CODE REVIEW

File: var/spack/repos/builtin/packages/fabulous/package.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    # cmake's specific
+    variant("shared", default=True, description="Build as shared library")
+
+    # fabulous's specific
+    variant("examples", default=False, description="build examples and tests")
+    variant("blasmt", default=False, description="use multi-threaded blas and lapack kernels")
+
+    # Dependencies
+    depends_on("pkg-config", type='build')
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/fabulous/package.py]: Is this a hard dependency on `pkg-config` or would the `pkgconfig` virtual work too (so `pkgconf` could be picked up when appropriate)?

File: var/spack/repos/builtin/packages/fabulous/package.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/fabulous/package.py]: Optional:  Would prefer the copyright end dat of `2024` in these packages but the checks will allow 2023 for the time being.

File: var/spack/repos/builtin/packages/maphyspp/package.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    maintainers("fpruvost")
+
+    version('1.1.9', '472deef05f69c26337a6f8e769cf36cbe0a50e6ec096757389ed10286a0d7e04',
+            url='https://gitlab.inria.fr/api/v4/projects/6194/packages/generic/source/v1.1.9/maphyspp-1.1.9.tar.gz')
+    version("master", branch="master",  submodules=True)
+
+    # User options
+    variant("armadillo", default=False, description="Compile examples and tests using Armadillo")
+    variant("pastix", default=True,  description="Compile examples and tests using pastix (sparse direct solver)")
+    #variant("mumps", default=False, description="Compile examples and tests using mumps (sparse direct solver)") #FIXME
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/maphyspp/package.py]: Do you plan to fix these commented out variants before or after the merge?  (Style checks will likely complain about the lack of a space after the comment marks.)

File: var/spack/repos/builtin/packages/fabulous/package.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            # and COMMONLY DEPLOYED (Fixed in Lapack 3.7.1 (25 June 2017))
+            "-DFABULOUS_LAPACKE_NANCHECK=OFF",
+            "-DFABULOUS_USE_CHAMELEON=OFF",
+            "-DCMAKE_BUILD_TYPE=%s"    % ('Debug'  if '+debug'    in spec else 'Release'),
+            "-DFABULOUS_DEBUG_MODE=%s" % ('ON'     if '+debug'    in spec else 'OFF'),
+            "-DBUILD_SHARED_LIBS=%s"   % ('ON'     if '+shared'   in spec else 'OFF'),
+            "-DFABULOUS_BLASMT=%s"     % ('ON'     if '+blasmt'   in spec else 'OFF'),
+            "-DFABULOUS_BUILD_EXAMPLES=%s" % ('ON' if '+examples' in spec else 'OFF'),
+            "-DFABULOUS_BUILD_TESTS=%s"    % ('ON' if '+examples' in spec else 'OFF'),
+        ])
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/fabulous/package.py]: There are helper methods you can use for `CMakePackage`s, such as `self.define_from_variant()` (see the docs at https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#adding-flags-to-cmake).

**Review by tldahlgren — COMMENTED** [review]
Confirmed the homepages, branches, and version directive's sha256s.

File: var/spack/repos/builtin/packages/fabulous/package.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+            "-DFABULOUS_BUILD_C_API=ON",
+            "-DFABULOUS_BUILD_Fortran_API=ON",
+
+            # This must stay like this until tpmqrt lapacke iface
+            # workspace bug is fixed in lapacke releases
+            # and COMMONLY DEPLOYED (Fixed in Lapack 3.7.1 (25 June 2017))
+            "-DFABULOUS_LAPACKE_NANCHECK=OFF",
+            "-DFABULOUS_USE_CHAMELEON=OFF",
+            "-DCMAKE_BUILD_TYPE=%s"    % ('Debug'  if '+debug'    in spec else 'Release'),
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/fabulous/package.py]: I'm not seeing a `debug` variant in this package.

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/fabulous/package.py]: Should this be coming from the `build_type`, where the options are:

```
$ spack info fabulous
...
    when build_system=cmake
      build_type [Release]      Debug, MinSizeRel, RelWithDebInfo, Release
          CMake build type
...
```

File: var/spack/repos/builtin/packages/paddle/package.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        args = []
+
+        args.extend([
+            "-Wno-dev",
+            "-DCMAKE_COLOR_MAKEFILE:BOOL=ON",
+            "-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON",
+            "-DCMAKE_BUILD_TYPE=%s"         % ('Debug' if '+debug'   in spec else 'Release'),
+            "-DBUILD_SHARED_LIBS=%s"        % ('ON'    if '+shared'  in spec else 'OFF'),
+            "-DPADDLE_BUILD_TESTS=%s"       % ('ON'    if '+tests'   in spec else 'OFF'),
+            "-DPADDLE_ORDERING_PARMETIS=%s" % ('ON'    if '+parmets' in spec else 'OFF')
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/paddle/package.py]: Typo, but I do recommend using the helper method for this.

```suggestion
            "-DPADDLE_ORDERING_PARMETIS=%s" % ('ON'    if '+parmetis' in spec else 'OFF')
```

File: var/spack/repos/builtin/packages/fabulous/package.py (hunk 5 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/fabulous/package.py]: You might want to constrain the build types if you only support `Debug` and `Release`.

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Going to switch to request changes since a typo needs to be fixed at a minimum.

**Comment by tldahlgren** [comment]
Note I also checked `spack spec maphyspp` and it succeeded.

**Comment by fpruvost** [comment]
Hi, thanks for your review. 
I pushed a new version, in a single commit, trying to clean a maximum.
I see there are some style errors in the ci, should I submit a new version with black / spack style polishes ?

**Comment by fpruvost** [comment]
Ok done, I pushed a new version with black applied.

**Review by tldahlgren — COMMENTED** [review]
LGTM.  

I suspect I forgot to mentioned the convention for listing branch version directives before numbered ones.  You can see an example at https://spack.readthedocs.io/en/latest/packaging_guide.html#fetching-from-code-repositories.

I'm okay with merging this PR as is if you would prefer to defer the change to another PR.  Just let me know.

**Comment by fpruvost** [comment]
> I'm okay with merging this PR as is if you would prefer to defer the change to another PR. Just let me know.

Ok done, I pushed an updated version with branch version before numbered ones.

**Review by tldahlgren — APPROVED** [review]
Thanks for the updates!

**Pull request merged**